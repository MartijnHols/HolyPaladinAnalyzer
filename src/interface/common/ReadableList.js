import React from 'react';
import PropTypes from 'prop-types';

/**
 * I think this is ugly but it works so I am going to concede and use it. Turns React elements into a human readable sentence.
 * E.g. <Maintainer1><Maintainer2><Maintainer3> becomes <Maintainer1>, <Maintainer2> and <Maintainer3>
 */
const ReadableList = ({ children, groupType }) => {
  const numItems = children.length;
  const results = [];
  React.Children.forEach(children, (child, index) => {
    const isFirst = index === 0;
    if (!isFirst) {
      const isLast = index === numItems - 1;
      if (isLast) {
        results.push(` ${groupType} `);
      } else {
        results.push(', ');
      }
    }
    results.push(child);
  });
  return results;
};
ReadableList.propTypes = {
  children: PropTypes.node.isRequired,
  groupType: PropTypes.oneOf(['and', 'or']),
};
ReadableList.defaultProps = {
  groupType: 'and',
};

export default ReadableList;
