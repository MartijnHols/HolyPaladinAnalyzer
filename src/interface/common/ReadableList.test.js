import React from 'react';
import { shallow } from 'enzyme';

import ReadableList from './ReadableList';

const Dummy = () => <div />;

describe('ReadableList', () => {
  it('doesn\'t mutate a single item', () => {
    expect(shallow(<ReadableList><Dummy /></ReadableList>)).toMatchSnapshot();
  });
  it('merges two items with an "and"', () => {
    expect(shallow(<ReadableList><Dummy /><Dummy /></ReadableList>)).toMatchSnapshot();
  });
  it('merges three items with a "," and an "and"', () => {
    expect(shallow(<ReadableList><Dummy /><Dummy /><Dummy /></ReadableList>)).toMatchSnapshot();
  });
  it('merges four items with two "," and an "and"', () => {
    expect(shallow(<ReadableList><Dummy /><Dummy /><Dummy /><Dummy /></ReadableList>)).toMatchSnapshot();
  });
});
