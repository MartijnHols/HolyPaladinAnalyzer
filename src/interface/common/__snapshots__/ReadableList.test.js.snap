// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadableList doesn't mutate a single item 1`] = `<Dummy />`;

exports[`ReadableList merges four items with two "," and an "and" 1`] = `
Array [
  <Dummy />,
  ", ",
  <Dummy />,
  ", ",
  <Dummy />,
  " and ",
  <Dummy />,
]
`;

exports[`ReadableList merges three items with a "," and an "and" 1`] = `
Array [
  <Dummy />,
  ", ",
  <Dummy />,
  " and ",
  <Dummy />,
]
`;

exports[`ReadableList merges two items with an "and" 1`] = `
Array [
  <Dummy />,
  " and ",
  <Dummy />,
]
`;
